### PR TITLE
CART-688 uri: PSR and lookup changes, stage1

### PR DESCRIPTION
**Pull request by frostedcmos**
URI lookup logic modified. Current model:

- For clients, when attempting to lookup uri for rank:tag=x,
   rank:tag=0 is first checked in the cache. If rank:tag=0 is present, it is used as
  a target of URI lookup request for rank:tag=x. Otherwise psr_rank:tag=0
  is used as a target for URI lookup request.

NOTE: This logic will further be modified/expanded in full CART-688 implementation.

- Servers are assumed to always know rank:tag=0 URIs for any rank
  as tag=0 URI is what servers primary ranks are added with.
  As such, servers will request rank:tag=x from rank:tag=0

- URI lookup handler on the server side will either return URI
  for rank:tag=x or if not found (psr might not know all tag URIs of all servers)
  it will return rank:tag=0 URI back. rank:tag=0 URI is guaranteed to be known
  if the rank is part of the group; otherwise -DER_OOG is returned to the caller.

  ul_tag field of the output of the URI_LOOKUP request will indicate which tag the
  uri was returned for.

As such for clients, when clients attempt to lookup URI of rank:tag=x
for the PSR:

- If a lookup request is made for rank:tag=x and instead rank:tag=0 is
  returned by the server (as indicated by ul_tag output field), then
  rank:tag=0 URI will be populated in the local cache and new URI request
  going to rank:tag=0 will be issued to lookup rank:tag=x

Other changes:
- Internal and FI RPC opcodes can be now looked up to print string
  name using crt_get_rpc_name() internal helper function.

Signed-off-by: Alexander Oganezov <alexander.a.oganezov@intel.com>

### CODE REVIEW

File: src/cart/src/cart/crt_rpc.h [hunk 1]
```diff
@@ -314,6 +316,7 @@ CRT_GEN_STRUCT(crt_grp_cache, CRT_SEQ_GRP_CACHE)
 
 #define CRT_OSEQ_URI_LOOKUP	/* output fields */		 \
 	((crt_phy_addr_t)	(ul_uri)		CRT_VAR) \
+	((uint32_t)		(ul_tag)		CRT_VAR) \
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.h:319]: Technically this is a protocol change so we should bump the number.

Comment (frostedcmos) [reply src/cart/src/cart/crt_rpc.h:319]: good point, will do

File: src/cart/src/cart/crt_group.c [hunk 2]
```diff
@@ -1638,44 +1539,33 @@ crt_hdlr_uri_lookup(crt_rpc_t *rpc_req)
...
 		D_GOTO(out, rc);
 
+	/* If this server does not know rank:0 then return error */
+	if (ul_in->ul_tag == 0)
+		D_GOTO(out, rc = -DER_OOG);
+
 	/**
-	 * step 2, make sure URI of tag 0 is in cache in preparation to forward
-	 * the lookup RPC
+	 * step 2, if rank:tag is not found, looup rank:tag=0
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1552]: (style)  trailing whitespace

File: src/cart/src/cart/crt_group.c [hunk 3]
```diff
@@ -1638,44 +1539,33 @@ crt_hdlr_uri_lookup(crt_rpc_t *rpc_req)
...
-		if (rc != 0) {
-			D_ERROR("crt_grp_lc_lookup(grp %s, rank %d, tag %d) "
-				"failed, rc: %d.\n",
-				grp_priv_primary->gp_pub.cg_grpid,
-				g_rank, 0, rc);
-			D_GOTO(out, rc);
-		}
+	if (rc != 0) {
+		D_ERROR("crt_grp_lc_lookup(grp %s, rank %d, tag %d) "
+			"failed, rc: %d.\n",
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1559]: (style)  quoted string split across lines

File: src/cart/src/cart/crt_group.c [hunk 4]
```diff
@@ -1638,44 +1539,33 @@ crt_hdlr_uri_lookup(crt_rpc_t *rpc_req)
...
-
-	/* step 3, forward the request to the final target */
-	rc = crt_uri_lookup_forward(rpc_req, g_rank);
-	if (rc != 0)
-		D_ERROR("crt_uri_lookup_forward() failed, rc %d\n", rc);
-
-	if (should_decref)
-		crt_grp_priv_decref(grp_priv);
-	return;
+
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1568]: (style)  trailing whitespace

File: src/cart/src/cart/crt_rpc.c [hunk 5]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
-crt_req_uri_lookup_by_rpc_cb(const struct crt_cb_info *cb_info)
+uri_lookup_cb(const struct crt_cb_info *cb_info)
 {
-	d_rank_t			 rank;
-	uint32_t			 tag;
-	crt_endpoint_t			*tgt_ep;
-	struct crt_rpc_priv		*rpc_priv;
-	struct crt_grp_priv		*grp_priv;
-	struct crt_context		*crt_ctx;
+	struct crt_rpc_priv 		*chained_rpc_priv;
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:548]: (style)  please, no space before tabs

File: src/cart/src/cart/crt_rpc.c [hunk 6]
```diff
@@ -811,6 +828,23 @@ crt_req_ep_lc_lookup(struct crt_rpc_priv *rpc_priv, bool *uri_exists)
 		D_GOTO(out, rc);
 	}
 
+	if (base_addr == NULL && rpc_priv->crp_hg_addr == NULL) {
+		if (crt_req_is_self(rpc_priv)) {
+			rc = crt_self_uri_get(tgt_ep->ep_tag, &uri);
+			if (rc != DER_SUCCESS) {
+				D_ERROR("crt_self_uri_get(tag: %d) failed, "
+				"rc %d\n", tgt_ep->ep_tag, rc);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:836]: (style)  quoted string split across lines

File: src/cart/src/cart/crt_group.c (hunk 2 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1552]: (style)  trailing whitespace

File: src/cart/src/cart/crt_group.c (hunk 3 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1559]: (style)  quoted string split across lines

File: src/cart/src/cart/crt_group.c (hunk 4 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1568]: (style)  trailing whitespace

File: src/cart/src/cart/crt_group.c (hunk 2 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1552]:
```suggestion
	 * step 2, if rank:tag is not found, looup rank:tag=0
```

File: src/cart/src/cart/crt_rpc.c (hunk 5 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:548]: (style)  please, no space before tabs

File: src/cart/src/cart/crt_rpc.c (hunk 6 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:836]: (style)  quoted string split across lines

File: src/cart/src/cart/crt_rpc.c [hunk 7]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
-
-	rpc_priv->crp_state = RPC_STATE_INITED;
-	rc = crt_req_send_internal(rpc_priv);
-
-out:
-	return rc;
-}
+static int
+crt_issue_uri_lookup(crt_context_t ctx, crt_group_t *group,
+			d_rank_t contact_rank, uint32_t contact_tag,
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:541]:
```suggestion
		     d_rank_t contact_rank, uint32_t contact_tag,
```

File: src/cart/src/cart/crt_rpc.c (hunk 5 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:548]:
```suggestion
	struct crt_rpc_priv		*chained_rpc_priv;
```

File: src/cart/src/cart/crt_rpc.c [hunk 8]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
+	}
 
-	uri = ul_out->ul_uri;
+	grp_priv = chained_rpc_priv->crp_grp_priv;
+	ctx = lookup_rpc->cr_ctx;
 
-	/* insert uri to hash table */
-	rc = crt_grp_lc_uri_insert(grp_priv, crt_ctx->cc_idx, rank, tag, uri);
+	rc = crt_grp_lc_uri_insert(grp_priv, ctx->cc_idx,
+			ul_in->ul_rank, ul_out->ul_tag, ul_out->ul_uri);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:577]:
```suggestion
				   ul_in->ul_rank, ul_out->ul_tag, ul_out->ul_uri);
```

File: src/cart/src/cart/crt_rpc.c [hunk 9]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
+	/* Lookup request will either return tag=ul_in->ul_tag URI or
+	 * tag=0 URI if ul_in->ul_tag is not found in server-side cache.
+	 *
+	 * ul_out->ul_tag will point to which tag URI was returned for.
+	 * If requested tag does not match returned tag, issue URI
+	 * request directly to the rank:tag=0 server
+	 */
+	if (ul_in->ul_tag != ul_out->ul_tag) {
+		rc = crt_issue_uri_lookup(lookup_rpc->cr_ctx,
+					lookup_rpc->cr_ep.ep_grp,
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:593]:
```suggestion
					  lookup_rpc->cr_ep.ep_grp,
```

File: src/cart/src/cart/crt_rpc.c [hunk 10]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
 		D_GOTO(out, rc);
 	}
 
-	rc = crt_req_send_internal(rpc_priv);
+	rc = crt_req_fill_tgt_uri(chained_rpc_priv, ul_out->ul_uri);
 	if (rc != 0) {
-		D_ERROR("crt_req_send_internal() failed, rc %d, opc: %#x\n",
-			rc, rpc_priv->crp_pub.cr_opc);
+		RPC_ERROR(chained_rpc_priv,
+			"crt_req_fill_tgt_uri() failed; rc=%d\n", rc);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:603]:
```suggestion
			  "crt_req_fill_tgt_uri() failed; rc=%d\n", rc);
```

File: src/cart/src/cart/crt_rpc.c [hunk 11]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
+	RPC_DECREF(chained_rpc_priv);
+}
+
+/**
+ * Helper function that returns rank for clients to contact
+ * for performing uri lookups
+ */
+static d_rank_t
+crt_client_get_contact_rank(crt_context_t crt_ctx, crt_group_t *grp,
+			d_rank_t query_rank, uint32_t query_tag)
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:631]:
```suggestion
			    d_rank_t query_rank, uint32_t query_tag)
```

File: src/cart/src/cart/crt_rpc.c [hunk 12]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
+	struct crt_context	*ctx;
+	int			rc;
+
+	grp_priv = crt_grp_pub2priv(grp);
+	ctx = crt_ctx;
+
+	/* If query_rank:tag=0 is in cache, use it as contact destination */
+	if (query_tag != 0) {
+		rc = crt_grp_lc_lookup(grp_priv, ctx->cc_idx,
+				query_rank, 0, &cached_uri, NULL);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:645]:
```suggestion
				       query_rank, 0, &cached_uri, NULL);
```

File: src/cart/src/cart/crt_rpc.c [hunk 13]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
 
-	ul_tgt_ep.ep_grp = orig_tgt_ep->ep_grp;
-	ul_tgt_ep.ep_rank = rank;
-	ul_tgt_ep.ep_tag = tag;
+	/* Client handling */
+	if (!crt_is_service()) {
+		d_rank_t lookup_rank;
+
+		lookup_rank = crt_client_get_contact_rank(ctx, grp,
+						tgt_ep->ep_rank,
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:676]:
```suggestion
							  tgt_ep->ep_rank,
```

File: src/cart/src/cart/crt_rpc.c [hunk 14]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
+			D_GOTO(out, rc = -DER_NONEXIST);
+		}
 
-	rc = crt_req_create(rpc_priv->crp_pub.cr_ctx, &ul_tgt_ep,
-			    CRT_OPC_URI_LOOKUP, &ul_req);
-	if (rc != 0) {
-		D_ERROR("crt_req_create URI_LOOKUP failed, rc: %d opc: %#x.\n",
-			rc, rpc_priv->crp_pub.cr_opc);
+		rc = crt_issue_uri_lookup(ctx, grp,
+					lookup_rank, 0,
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:684]:
```suggestion
					  lookup_rank, 0,
```

File: src/cart/src/cart/crt_rpc.c [hunk 15]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
 
-		/* Errors will trigger complete_cb that decrefs
-		 * addrefs taken above.
-		 */
+	/* Server handling */
+
+	/* Servers must know tag=0 uris of other servers */
+	if (tgt_ep->ep_tag == 0) {
+		RPC_ERROR(rpc_priv, "Target %d:%d not known\n",
+			tgt_ep->ep_rank, tgt_ep->ep_tag);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:695]:
```suggestion
			  tgt_ep->ep_rank, tgt_ep->ep_tag);
```

File: src/cart/src/cart/crt_rpc.c [hunk 16]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
+		D_GOTO(out, rc = -DER_OOG);
 	}
 
+	/* Send request to tag=0 to get uri for ep_tag */
+	rc = crt_issue_uri_lookup(ctx, grp,
+			tgt_ep->ep_rank, 0,
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:701]:
```suggestion
				  tgt_ep->ep_rank, 0,
```

File: src/cart/src/cart/crt_rpc.c [hunk 17]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
-/*
- * Contact the PSR to obtain the URI of the target rank which rpc_priv is trying
- * to communicate to
- */
-int
-crt_req_uri_lookup_psr(struct crt_rpc_priv *rpc_priv, crt_cb_t complete_cb,
-		       void *arg)
+static int
+crt_issue_uri_lookup(crt_context_t ctx, crt_group_t *group,
+			d_rank_t contact_rank, uint32_t contact_tag,
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:710]:
```suggestion
		     d_rank_t contact_rank, uint32_t contact_tag,
```

File: src/cart/src/cart/crt_rpc.c [hunk 18]
```diff
@@ -811,6 +828,23 @@ crt_req_ep_lc_lookup(struct crt_rpc_priv *rpc_priv, bool *uri_exists)
...
+		if (crt_req_is_self(rpc_priv)) {
+			rc = crt_self_uri_get(tgt_ep->ep_tag, &uri);
+			if (rc != DER_SUCCESS) {
+				D_ERROR("crt_self_uri_get(tag: %d) failed, "
+				"rc %d\n", tgt_ep->ep_tag, rc);
+				D_GOTO(out, rc);
+			}
+
+			rc = crt_grp_lc_uri_insert(grp_priv, ctx->cc_idx,
+				tgt_ep->ep_rank, tgt_ep->ep_tag, base_addr);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:841]:
```suggestion
						   tgt_ep->ep_rank, tgt_ep->ep_tag, base_addr);
```

File: src/cart/src/cart/crt_rpc.h [hunk 19]
```diff
@@ -314,6 +314,7 @@ CRT_GEN_STRUCT(crt_grp_cache, CRT_SEQ_GRP_CACHE)
...
 	((int32_t)		(ul_rc)			CRT_VAR)
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.h:318]:
```suggestion
	((crt_phy_addr_t)(ul_uri)		CRT_VAR) \
	((uint32_t)(ul_tag)		CRT_VAR) \
	((int32_t)(ul_rc)			CRT_VAR)
```

**Comment by ashleypittman** [comment]
I've managed to run the checkpatch stage locally and have identified the issue, the problem is it's asking for changes to the RPC definitions crt_rpc.h and there's lots of them close together so checkpatch is merging them into one but github isn't.  Working on resolving it now.

File: src/cart/src/cart/crt_group.c (hunk 2 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1552]: (style)  trailing whitespace

File: src/cart/src/cart/crt_group.c (hunk 3 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1559]: (style)  quoted string split across lines

File: src/cart/src/cart/crt_group.c (hunk 4 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1568]: (style)  trailing whitespace

File: src/cart/src/cart/crt_group.c (hunk 2 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1552]:
```suggestion
	 * step 2, if rank:tag is not found, looup rank:tag=0
```

File: src/cart/src/cart/crt_rpc.c (hunk 5 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:548]: (style)  please, no space before tabs

File: src/cart/src/cart/crt_rpc.c (hunk 6 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:836]: (style)  quoted string split across lines

File: src/cart/src/cart/crt_rpc.c [hunk 20]
```diff
@@ -536,224 +536,222 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
...
+	 * If requested tag does not match returned tag, issue URI
+	 * request directly to the rank:tag=0 server
+	 */
+	if (ul_in->ul_tag != ul_out->ul_tag) {
+		rc = crt_issue_uri_lookup(lookup_rpc->cr_ctx,
+					lookup_rpc->cr_ep.ep_grp,
+					ul_in->ul_rank, 0,
+					ul_in->ul_rank, ul_in->ul_tag,
+					chained_rpc_priv);
 		D_GOTO(out, rc);
```
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:597]:
```suggestion
	D_ALLOC(tmp_array, sizeof(*tmp_array) * count);
	if (tmp_array == NULL)
		D_GOTO(out, rc = -DER_NOMEM);
	memcpy(tmp_array, ver, sizeof(tmp_array[0]) * count);
```

File: src/cart/src/cart/crt_rpc.c (hunk 7 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:541]:
```suggestion
		     d_rank_t contact_rank, uint32_t contact_tag,
```

File: src/cart/src/cart/crt_rpc.c (hunk 5 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:548]:
```suggestion
	struct crt_rpc_priv		*chained_rpc_priv;
```

File: src/cart/src/cart/crt_rpc.c (hunk 8 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:577]:
```suggestion
				   ul_in->ul_rank, ul_out->ul_tag, ul_out->ul_uri);
```

File: src/cart/src/cart/crt_rpc.c (hunk 9 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:593]:
```suggestion
					  lookup_rpc->cr_ep.ep_grp,
```

File: src/cart/src/cart/crt_rpc.c (hunk 10 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:603]:
```suggestion
			  "crt_req_fill_tgt_uri() failed; rc=%d\n", rc);
```

File: src/cart/src/cart/crt_rpc.c (hunk 11 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:631]:
```suggestion
			    d_rank_t query_rank, uint32_t query_tag)
```

File: src/cart/src/cart/crt_rpc.c (hunk 12 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:645]:
```suggestion
				       query_rank, 0, &cached_uri, NULL);
```

File: src/cart/src/cart/crt_rpc.c (hunk 13 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:676]:
```suggestion
							  tgt_ep->ep_rank,
```

File: src/cart/src/cart/crt_rpc.c (hunk 14 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:684]:
```suggestion
					  lookup_rank, 0,
```

File: src/cart/src/cart/crt_rpc.c (hunk 15 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:695]:
```suggestion
			  tgt_ep->ep_rank, tgt_ep->ep_tag);
```

File: src/cart/src/cart/crt_rpc.c (hunk 16 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:701]:
```suggestion
				  tgt_ep->ep_rank, 0,
```

File: src/cart/src/cart/crt_rpc.c (hunk 17 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:710]:
```suggestion
		     d_rank_t contact_rank, uint32_t contact_tag,
```

File: src/cart/src/cart/crt_rpc.c (hunk 18 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:841]:
```suggestion
						   tgt_ep->ep_rank, tgt_ep->ep_tag, base_addr);
```

File: src/cart/src/cart/crt_rpc.h (hunk 19 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.h:318]:
```suggestion
	((crt_group_id_t)(ul_grp_id)		CRT_VAR) \
	((d_rank_t)(ul_rank)		CRT_VAR) \
	((uint32_t)(ul_tag)		CRT_VAR)

#define CRT_OSEQ_URI_LOOKUP	/* output fields */		 \
	((crt_phy_addr_t)(ul_uri)		CRT_VAR) \
	((uint32_t)(ul_tag)		CRT_VAR) \
	((int32_t)(ul_rc)			CRT_VAR)
```

File: src/cart/src/cart/crt_group.c (hunk 2 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1552]: Spelling.

Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1552]: (style)  trailing whitespace

File: src/cart/src/cart/crt_group.c (hunk 3 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1559]: (style)  quoted string split across lines

File: src/cart/src/cart/crt_group.c (hunk 4 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_group.c:1568]: (style)  trailing whitespace

File: src/cart/src/cart/crt_rpc.c (hunk 5 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:548]: (style)  please, no space before tabs

File: src/cart/src/cart/crt_rpc.c (hunk 6 above)
Comment (ashleypittman) [inline src/cart/src/cart/crt_rpc.c:836]: (style)  quoted string split across lines

File: src/cart/src/cart/crt_group.c (hunk 2 above)
Comment (frostedcmos) [reply src/cart/src/cart/crt_group.c]: fixed

**Comment by frostedcmos** [comment]
closing as git merge has screwed up

**Pull request closed without merging**